src/control/server/mgmt_svc.go

```diff
@@ -530,110 +530,129 @@ func (svc *mgmtSvc) PoolDeleteACL(ctx context.Context, req *mgmtpb.DeleteACLReq)
 func (svc *mgmtSvc) BioHealthQuery(ctx context.Context, req *mgmtpb.BioHealthReq) (*mgmtpb.BioHealthResp, error) {
 	svc.log.Debugf("MgmtSvc.BioHealthQuery dispatch, req:%+v\n", *req)
 
-	mi, err := svc.harness.GetMSLeaderInstance()
-	if err != nil {
-		return nil, err
-	}
+	for _, i := range svc.harness.Instances() {
+		dresp, err := i.CallDrpc(drpc.ModuleMgmt, drpc.MethodBioHealth, req)
+		if err != nil {
+			continue
+		}
 
-	dresp, err := mi.CallDrpc(drpc.ModuleMgmt, drpc.MethodBioHealth, req)
-	if err != nil {
-		return nil, err
-	}
+		resp := &mgmtpb.BioHealthResp{}
+		if err = proto.Unmarshal(dresp.Body, resp); err != nil {
+			return nil, errors.Wrap(err, "unmarshal BioHealthQuery response")
+		}
 
-	resp := &mgmtpb.BioHealthResp{}
-	if err = proto.Unmarshal(dresp.Body, resp); err != nil {
-		return nil, errors.Wrap(err, "unmarshal BioHealthQuery response")
+		if resp.Status == 0 {
+			return resp, nil
+		}
 	}
 
-	return resp, nil
+	reqID := func() string {
+		if req.DevUuid != "" {
+			return req.DevUuid
+		}
+		return req.TgtId
+	}
+	return nil, errors.Errorf("no rank matched request for %q", reqID())
 }
 
 // SmdListDevs implements the method defined for the Management Service.
 func (svc *mgmtSvc) SmdListDevs(ctx context.Context, req *mgmtpb.SmdDevReq) (*mgmtpb.SmdDevResp, error) {
 	svc.log.Debugf("MgmtSvc.SmdListDevs dispatch, req:%+v\n", *req)
 
-	mi, err := svc.harness.GetMSLeaderInstance()
-	if err != nil {
-		return nil, err
-	}
+	fullResp := new(mgmtpb.SmdDevResp)
 
-	dresp, err := mi.CallDrpc(drpc.ModuleMgmt, drpc.MethodSmdDevs, req)
-	if err != nil {
-		return nil, err
-	}
+	for _, i := range svc.harness.Instances() {
+		dresp, err := i.CallDrpc(drpc.ModuleMgmt, drpc.MethodSmdDevs, req)
+		if err != nil {
+			return nil, err
+		}
 
-	resp := &mgmtpb.SmdDevResp{}
-	if err = proto.Unmarshal(dresp.Body, resp); err != nil {
-		return nil, errors.Wrap(err, "unmarshal SmdListDevs response")
+		tmpResp := new(mgmtpb.SmdDevResp)
+		if err = proto.Unmarshal(dresp.Body, tmpResp); err != nil {
+			return nil, errors.Wrap(err, "unmarshal SmdListDevs response")
+		}
+
+		if tmpResp.Status != 0 {
+			return tmpResp, nil
+		}
```

Comment:
Not necessary to address right now, but I wonder if there might be some use in reporting errors on a per-instance basis. I.e. is it possible one could return error, while the other returns a valid list of devices? Seems like both pieces of information could be useful to the caller, especially if they're trying to debug.

Reply:
Yeah, I'm trying to keep the scope here as narrow as possible for now. I agree that this would probably be better. https://jira.hpdd.intel.com/browse/DAOS-3959